iroh/src/magicsock.rs

```diff
@@ -1561,7 +1633,50 @@ enum DiscoBoxError {
     Parse(anyhow::Error),
 }
 
-type RelayRecvResult = Result<(PublicKey, quinn_udp::RecvMeta, Bytes), io::Error>;
+/// Channel for [`MagicSock::poll_recv_relay`] to receive QUIC datagrams from relays.
+///
+/// This channel contains both the send and receive sides.  The receive side implements the
+/// poll interface with required wakers.
+#[derive(Debug)]
+struct RelayRecvChannel {
+    sender: mpsc::Sender<RelayReadResult>,
+    receiver: parking_lot::Mutex<mpsc::Receiver<RelayReadResult>>,
```

Comment:
Heh, that did cross my mind but it was so much simpler to put it in a single struct :smile:.  I was not expecting such aversion to this.

The nice thing about version is that a class of errors go away because there is always a sender.  But sure, if the sender starts calling recv and the other way round things would be... odd.  But also I didn't see that as a particular big risk.

Neither side of this need to be clonable.  I kind of like this as a newtype regardless of how we want to change the implementation.  But anyway, see my comment above on avoiding changing the implementation so far.